New PR for property/funciton conflict of 'domain'

kolmanthomas opened the pull request:
Okay I cleaned up the pull request and targeted it to octave_dev25. This PR only deals with the property/function conflict of domain, I will create a separate PR for dealing with array-valued chebfuns.

kolmanthomas committed:
Temp fix for 2*chebfun() to work

kolmanthomas committed:
Workaround for subsref issue with domain

cbm755 commented on @chebfun/mtimes.m:
```diff
@@ -88,7 +88,9 @@
-        f = mtimes(g.', f.').';
+        %f = mtimes(g.', f.').';
+        % Transposing g and f here is what causes 2*f to fail and f*2 to work.
+        f = mtimes(g, f).';
```
Maybe we can put "Octave specific workaround" into this comment?

kolmanthomas committed:
Add method to check if Octave is being used

kolmanthomas committed:
Revert "Temp fix for 2*chebfun() to work"

This reverts commit b5a9a93fcfde2deb73f41e2029db5c138a7bac72.

It wasn't the correct action to fix the 2*f issue, broke a lot of other behaviours like 'f + 2*f' et cetera.

kolmanthomas committed:
Revert "Workaround for subsref issue with domain"

This reverts commit b1165b502189f57ef0b92a9e30d17e009752b23c.

kolmanthomas committed:
Revert "octave: workaround subsref and property fighting over "domain""

This reverts commit 45a8b00f67e740aa0ebe73ba4c375ab5bfa7b55e.

kolmanthomas committed:
Added comma delimiter for cell array of anon functions.

See bug #67348 (https://savannah.gnu.org/bugs/?67348) for more details.

kolmanthomas committed:
Revert "octave: TEMPORARY dumb hack: always return [-1, 1] for domain"

This reverts commit d1c57258223b3efc5c65a54b8fdfb6c35ff6f8ca.

kolmanthomas committed:
Revert "octave: temporary workaround for f.funs{j}"

This reverts commit 22aac2545a3468ed478c0039250707793ead3039.

Issue was fixed in core Octave (hg id 34756:b12978fb1380) (see https://savannah.gnu.org/bugs/index.php?65179).

kolmanthomas committed:
Fix unintended behaviour making f a complex-valued chebfun.

test_atan2 passes without error after this change.

kolmanthomas committed:
Added a new function to handle 'verLessThan('matlab', version)' calls.

Previously, calling `verLessThan('matlab', version)` caused an error on Octave. With this change, querying the `matlab` version will silently query the Octave version on Octave. All invocations of `verLessThan` must be changed to `compatible_verLessThan`.

kolmanthomas committed:
Changed verLessThan to compatible_verLessThan

kolmanthomas committed:
Ignore these tests for now.

Octave does not support bvpinit() as of now.

kolmanthomas committed:
Ignore this test for now, it crashes the interpreter.

kolmanthomas committed:
Revert "octave: use static methods instead of class-related functions"

This reverts commit 0232943c64beb51363b957256b231712b0849196.

Methods inside a classdef .m file that do not belong to the class itself are now supported in Octave.

kolmanthomas committed:
Workaround for isequal precendence call.

Octave does not currently support the InferiorClasses attribute, so often two classes with the same method will conflict, and the order of function resolution will be messed up.

kolmanthomas committed:
Declare fun class as Abstract.

Currently, Octave does not declare the whole class to be abstract if a method or a property is abstract, unlike MATLAB, which does even for a single abstract method or property.

kolmanthomas committed:
Replaced 'fields' method with the 'fieldnames' method.

The former doesn't exist in Octave, nor can I find official MATLAB documentation for the 'fields' method, but it seems to do the exact same thing as the 'fieldnames' method.

kolmanthomas committed:
Changes to plot.m by changing 'verLessThan' calls to 'compatible_verLessThan'.

See commit 2a6a13620732632d4439c2e8e8d3342e3c25fbac for details on the reasoning.

kolmanthomas committed:
Renamed filename resetPointValues to clearPointValues in order to match function name.

Silences an Octave warning.

kolmanthomas committed:
Changes to seedRNG.m by changing 'verLessThan' calls to 'compatible_verLessThan'.

kolmanthomas commented:
Okay, I rebased on my own history and changed a few things. A couple of things:

1) 

You will need the bleeding-edge dev branch along with the patch from here (https://savannah.gnu.org/bugs/?67362) to get everything to work the best it can. The patch is very much needed, almost everything will break otherwise.

2) 

> Maybe we can put "Octave specific workaround" into this comment?

Not needed anymore, I reverted the change. I added similar comments in other commits where I saw fit, but some of them (like changes from `verLessThan` to`compatible_verLessThan` ) I thought didn't need comments.

3)

I decided against doing something like

`~is_octave && verLessThan('matlab', 'VER')`

because the logic would get hairy if I would have to start checking Octave versions as well. Ideally, I like having all the logic encapsulated in one function rather than spread out everywhere in separate files.

Instead, I propose

`compatible_verLessThan('matlab', 'VER')`

which, right now, returns `false` always when Octave is used. If we need to check the Octave version, we can modify the `compatible_verLessThan` method to accept something like

`compatible_verLessThan('matlab', 'VER', 'octave', 'VER')`

and it would check the Octave version when running Octave, and the MATLAB version when running MATLAB.

Let me know your thoughts on this approach, I'm not married to it and if there's a pitfall I'm not noticing, I can revert it.

IMO, the packaging of Chebfun as an Octave package can start now.

cbm755 commented:
Nice work!

I'm not sure I agree yet about `compatible_verLessThan` but this seems like a minor point that we can discuss elsewhere.

For now I'm going to merge this.
